pysnmp/proto/secmod/rfc3414/service.py

```diff
@@ -446,34 +446,40 @@ def __generateRequestOrResponseMsg(self, snmpEngine,
         )
 
         # 3.1.6a
-        if securityStateReference is None and securityLevel in (2, 3):
-            if securityEngineID in self.__timeline:
-                (snmpEngineBoots, snmpEngineTime, latestReceivedEngineTime,
+        snmpEngineBoots = snmpEngineTime = 0
+
+        if securityLevel in (2, 3):
+            pdu = scopedPDU.getComponentByPosition(2).getComponent()
+
+            # 3.1.6.b
+            if pdu.tagSet in rfc3411.unconfirmedClassPDUs:
+                (snmpEngineBoots,
+                 snmpEngineTime) = mibBuilder.importSymbols('__SNMP-FRAMEWORK-MIB', 'snmpEngineBoots', 'snmpEngineTime')
+
+                snmpEngineBoots = snmpEngineBoots.syntax
+                snmpEngineTime = snmpEngineTime.syntax.clone()
+
+                debug.logger & debug.flagSM and debug.logger(
+                    '__generateRequestOrResponseMsg: read snmpEngineBoots, snmpEngineTime from LCD')
+
+            # 3.1.6a
+            elif securityEngineID in self.__timeline:
+                (snmpEngineBoots,
+                 snmpEngineTime,
+                 latestReceivedEngineTime,
                  latestUpdateTimestamp) = self.__timeline[securityEngineID]
+
                 debug.logger & debug.flagSM and debug.logger(
                     '__generateRequestOrResponseMsg: read snmpEngineBoots, snmpEngineTime from timeline')
+
+            # 3.1.6.c
             else:
-                # 2.3 XXX is this correct?
-                snmpEngineBoots = snmpEngineTime = 0
                 debug.logger & debug.flagSM and debug.logger(
-                    f'__generateRequestOrResponseMsg: no timeline for securityEngineID {securityEngineID!r}')
-        # 3.1.6.b
-        elif securityStateReference is not None:  # XXX Report?
-            (snmpEngineBoots,
-             snmpEngineTime) = mibBuilder.importSymbols('__SNMP-FRAMEWORK-MIB', 'snmpEngineBoots', 'snmpEngineTime')
-            snmpEngineBoots = snmpEngineBoots.syntax
-            snmpEngineTime = snmpEngineTime.syntax.clone()
-            debug.logger & debug.flagSM and debug.logger(
-                '__generateRequestOrResponseMsg: read snmpEngineBoots, snmpEngineTime from LCD')
-        # 3.1.6.c
-        else:
-            snmpEngineBoots = snmpEngineTime = 0
-            debug.logger & debug.flagSM and debug.logger(
-                '__generateRequestOrResponseMsg: assuming zero snmpEngineBoots, snmpEngineTime')
+                    '__generateRequestOrResponseMsg: assuming zero snmpEngineBoots, snmpEngineTime')
 
-        debug.logger & debug.flagSM and debug.logger(
-            '__generateRequestOrResponseMsg: use snmpEngineBoots {} snmpEngineTime {} for securityEngineID {!r}'.format(
-                snmpEngineBoots, snmpEngineTime, securityEngineID))
+            debug.logger & debug.flagSM and debug.logger(
+                '__generateRequestOrResponseMsg: use snmpEngineBoots %s snmpEngineTime %s for securityEngineID %r' % (
```

Comment:
Ok, last thing is to make `pre-commit` step green. It requires running `pre-commit run --all-files` command in the root of the repository. It shows one error in this line. You can check the details here -> https://github.com/pysnmp/pysnmp/runs/6724646831?check_suite_focus=true

Reply:
Ouu... thanks, I'm checking it! :smile:

Reply:
Done ;)